policy/json_util.go

```diff
@@ -0,0 +1,143 @@
+package policy
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"time"
+)
+
+type MapOrder []string
+
+func (o *MapOrder) UnmarshalJSON(data []byte) error {
+	*o = retrieveKeyOrderingFromMarshaledMap(bytes.NewReader(data))
+	return nil
+}
+
+type listOrder struct {
+	PinsetOrder      MapOrder `json:"pinsets"`
+	PolicyAliasOrder MapOrder `json:"policy-aliases"`
+	PolicyOrder      MapOrder `json:"policies"`
+}
+
+// Like a list, but also stores metadata about orderings :)
+type OrderedList struct {
+	list
+	listOrder
+}
+
+type OrderedPinsetMap struct {
+	data map[string]Pinset
+	MapOrder
+}
+
+type OrderedPolicyMap struct {
+	data map[string]TLSPolicy
+	MapOrder
+}
+
+func (m OrderedPolicyMap) MarshalJSON() ([]byte, error) {
+	data := "{"
+	for _, key := range m.MapOrder {
+		value, err := json.Marshal(m.data[key])
+		if err != nil {
+			return nil, err
+		}
+		data += fmt.Sprintf("\"%s\":%s,", key, value)
+	}
+	return []byte(data[:len(data)-1] + "}"), nil
+}
+
+func (m OrderedPinsetMap) MarshalJSON() ([]byte, error) {
+	data := "{"
+	for _, key := range m.MapOrder {
+		value, err := json.Marshal(m.data[key])
+		if err != nil {
+			return nil, err
+		}
+		data += fmt.Sprintf("\"%s\":%s,", key, value)
+	}
+	return []byte(data[:len(data)-1] + "}"), nil
+
+}
```

Comment:
The worst part is these line-by-line duplicated methods. We should be able to define a common class, `OrderedMap`... but I couldn't figure out a nice way to do this. Let me know if you have any ideas here!